site/bps/BP-38-bookie-endpoint-discovery.md

```diff
@@ -0,0 +1,149 @@
+
+With this proposal we are now introducing a way for the Bookie to advertise the services it exposes, basically with this change the Bookie will be able to store on the Metadata Service a set of name-value pairs that describe the *available services*.
+
+We will also define a set of well know properties that will be useful for futher implementations.
+
+This information will be also useful for Monitoring and Management services as it will enable full discovery of the capabilities of all of the Bookies in a cluster just by having read access to the Metadata Service.
+
+### Public Interfaces
+
+On the Registration API, we introduce a new data structure that describes the services
+exposed by a Bookie:
+
+```
+inteface BookieServiceInfo {
+    class Endpoint {
+        String name;
+        String hostname;
+        int port;
+        String protocol; // "bookie-rpc", "http", "https"....
+        String[] auth; // "sasl-gsapi", "tls-client-auth", "sasl-md5"....
+        String[] extensions; // "starttls"
+    }
+    List<Endpoint> endpoints;
+    Map<String, String> properties; // additional properties
+}
+
+```
+
+In RegistrationClient interface we expose a new method:
+
+```
+CompletableFuture<Versioned<BookieServiceInfo>> getBookieServiceInfo(String bookieId)
+```
+
+The client can derive bookieId from a BookieSocketAddress. He can access the list of available bookies using **RegistrationClient#getAllBookies()** and then use this new method to get the details of the services exposed by the Bookie.
+
+On the Bookie side we change the RegistrationManager interface in order to let the Bookie
+advertise the services:
+
+in RegistrationManager class the **registerBookie** method 
+```
+void registerBookie(String bookieId, boolean readOnly)
+```
+
+becomes
+
+```
+void registerBookie(String bookieId, boolean readOnly, BookieServiceInfo bookieServieInfo)
+```
+
+It will be up to the implementation of RegistrationManager and RegistrationClient to serialize
+the BookieServiceInfo structure.
+
+For the ZooKeeper based implementation we are going to store such information in JSON format.
+
+```
+{
+    "endpoints": [
+        {
+         "name": "bookie",
+         "hostname": "localhost",
+         "port": 3181,
+         "protocol": "bookie-rpc",
+         "auth": ["sasl-gsapi"],
+         "extensions": ["starttls"]
+        },
+        {
+         "name": "bookie-http-server",
+         "hostname": "localhost",
+         "port": 8080,
+         "protocol": "http",
+         "auth": [],
+         "extensions": []
+        }
+    ],
+    properties {
+        "property": "value",
+        "property2": "value2"
+    }
+}
+```
+Such information will be stored inside the '/REGISTRATIONPATH/available' znode (or /REGISTRATIONPATH/available/readonline' in case of readonly bookie), these paths are the same used in 4.10, but in 4.10 we are writing empty z-nodes.
```

Comment:
I think you meant to say `/REGISTRATIONPATH/available/readonly`. Is it confusing to have two possible nodes storing it? Why don't you want it to always be in '/REGISTRATIONPATH/available'?

Reply:
Sorry for the typo.
We are already using these znodes:
`/REGISTRATIONPATH/available`
`/REGISTRATIONPATH/available/readonly`

I don't want to change this logic, it will make harder backward  compatibility

see:
https://github.com/apache/bookkeeper/blob/163db4496d50cae198b80c9d88c14579f6d2e9f3/bookkeeper-server/src/main/java/org/apache/bookkeeper/discover/ZKRegistrationClient.java#L188